discovery/manager.go

```diff
@@ -0,0 +1,293 @@
+	actionCh           chan func()
+	targetSetProviders map[string]map[string]*targetSetProvider // map[targetSetName]map[providerName]
+}
+
+// Run starts background processing
+func (m *DiscoveryManager) Run() error {
+	for {
+		select {
+		case f := <-m.actionCh:
+			f()
+		case <-m.ctx.Done():
+			return m.ctx.Err()
+		}
+	}
+
+}
+
+// SyncCh returns the read only channel to be used by another package to receive TargetGroup updates for a given target set
+func (m *DiscoveryManager) SyncCh() <-chan map[string][]*config.TargetGroup {
+	return m.syncCh
+}
+
+// ApplyConfig removes all running discovery providers and starts new ones using the provided config
+// discovery providers than start sending updates to the global sync channel
+func (m *DiscoveryManager) ApplyConfig(cfg *config.Config) error {
+	err := make(chan error)
+	m.actionCh <- func() {
+		m.cancelDiscoveryProviders()
+		for _, scfg := range cfg.ScrapeConfigs {
+			for provName, prov := range m.providersFromConfig(scfg.ServiceDiscoveryConfig) {
+				ctx, cancel := context.WithCancel(m.ctx)
+				updates := make(chan []*config.TargetGroup)
+
+				m.createProvider(cancel, scfg.JobName, provName)
+
+				go prov.Run(ctx, updates)
+				go func(provName string) {
+					select {
+					case <-ctx.Done():
+						// First set of all targets the provider knows.
+					case tgs, ok := <-updates:
+						// Handle the case that a target provider exits and closes the channel
+						// before the context is done.
+						if !ok {
+							break
+						}
+						m.syncCh <- m.mergeGroups(scfg.JobName, provName, tgs)
+					case <-time.After(5 * time.Second):
+						// Initial set didn't arrive. Act as if it was empty
+						// and wait for updates later on.
+					}
+
+					// Start listening for further updates.
+					for {
+						select {
+						case <-ctx.Done():
+							return
+						case tgs, ok := <-updates:
+							// Handle the case that a target provider exits and closes the channel
+							// before the context is done.
+							if !ok {
+								return
+							}
+							m.syncCh <- m.mergeGroups(scfg.JobName, provName, tgs)
+						}
+					}
+				}(provName)
+			}
+		}
+		close(err)
+	}
+
+	return <-err
+}
+
+func (m *DiscoveryManager) cancelDiscoveryProviders() {
+	for targetSetName, targetSetProviders := range m.targetSetProviders {
+		for _, discoveryProvider := range targetSetProviders {
+			discoveryProvider.cancel()
+		}
+		delete(m.targetSetProviders, targetSetName)
+	}
+}
+
+func (m *DiscoveryManager) createProvider(cancel context.CancelFunc, tsName, provName string) {
+	if m.targetSetProviders[tsName] == nil {
+		m.targetSetProviders[tsName] = make(map[string]*targetSetProvider)
+	}
+	m.targetSetProviders[tsName][provName] = &targetSetProvider{
+		cancel:  cancel,
+		tgroups: []*config.TargetGroup{},
+	}
+}
+
+// mergeGroups adds a new target group for a named discovery provider and returns all  target groups for a given target set
```

Comment:
Nitpick: Double space in comment here. Comments should also end with a period.

Reply:
thanks 
I will fix all comments and tests after I get one more OK from the core maintainers as it might require more changes and want to preserve some keystrokes :)